### PR TITLE
Add conn to jinja template context

### LINKED ISSUE #14597: Provide jinja template syntax to access connections
**Description**

Expose the connection into the jinja template context via `conn.value.<connectionname>.{host,port,login,password,extra_config,etc}`

Today is possible to conveniently access [airflow's variables](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#variables) in jinja templates using `{{ var.value.<variable_name> }}`. 

There is no equivalent (to my knowledge for [connections](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#connections)), I understand that most of the time connection are used programmatically in Operators and Hooks source code, but there are use cases where the connection info has to be pass as parameters to the operators and then it becomes cumbersome to do it without jinja template syntax.

I seen workarounds like using [user defined macros to provide get_login(my_conn_id)](https://stackoverflow.com/questions/65826404/use-airflow-connection-from-a-jinja-template/65873023#65873023
), but I'm after a consistent interface for accessing both variables and connections in the same way

**Workaround**
The following `user_defined_macro` (from my [stackoverflow answer](https://stackoverflow.com/a/66471911/90580)) provides the suggested syntax `connection.mssql.host` where `mssql` is the connection name: 

```
   class ConnectionGrabber:
       def __getattr__(self, name):
           return  Connection.get_connection_from_secrets(name)
dag = DAG( user_defined_macros={'connection': ConnectionGrabber()}, ...)
 task = BashOperator(task_id='read_connection', bash_command='echo {{connection.mssql.host }}', dag=dag)
```

This macro can be added to each DAG individually or to all DAGs via an Airflow's Plugin. What I suggest is to make this macro part of the default. 

**Use case / motivation**

For example, passing credentials to a [KubernetesPodOperator](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html#howto-operator-kubernetespodoperator) via [env_vars](https://cloud.google.com/composer/docs/how-to/using/using-kubernetes-pod-operator) today has to be done like this: 

```
connection = Connection.get_connection_from_secrets('somecredentials')
k = KubernetesPodOperator(
  task_id='task1', 
  env_vars={'MY_VALUE': '{{ var.value.my_value }}', 'PWD': conn.password,},
  )
```
where I would prefer to use consistent syntax for both variables and connections like this: 
```
# not needed anymore: connection = Connection.get_connection_from_secrets('somecredentials')
k = KubernetesPodOperator(
  task_id='task1', 
  env_vars={'MY_VALUE': '{{ var.value.my_value }}', 'PWD':  '{{ conn.somecredentials.password }}',},
  )
```

The same applies to `BashOperator` where I sometimes feel the need to pass connection information to the templated script.

**Are you willing to submit a PR?**

yes, I can write the PR.

**Related Issues**

### PR DESCRIPTION
**Pull request by ecerulm**
closes: #14597

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by ecerulm** [comment]
pinging @uranusjr , @potiuk , and @turbaszek  as they were involved in the discussion of #14597

**Comment by uranusjr** [comment]
Logic looks good to me, but this won’t pass static checks (formatting issues). Please try to run those locally and fix them first.

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]: It's not a variable name but a conn_id

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
            'conn': ConnectionAccessor(),
```

We don't need the `value` subitem here, as conn can only ever return one thing.

Comment (ecerulm) [reply airflow/models/taskinstance.py]: @ashb But don't we want consistency between the var accessor syntax and the connection accessor syntax? I mean, assuming people is already familiar with  syntax `var.value.my_connection` I guess they will try `conn.value.my_connection` and not `conn.my_connection`. 

If you already pondered about that and still think the `conn.my_connectio` is preferrable just tell me and I'll change it.

**Comment by ecerulm** [comment]
> Logic looks good to me, but this won’t pass static checks (formatting issues). Please try to run those locally and fix them first.

`pre-commit run` did not complain for me: 

```
git fetch origin
pre-commit run --from-ref $(git merge-base origin/main jinjaconn) --to-ref jinjaconn

identity.................................................................................Passed
- hook id: identity
- duration: 0.11s

airflow/models/taskinstance.py
tests/models/test_taskinstance.py

Check hooks apply to the repository..................................(no files to check)Skipped
No-tabs checker..........................................................................Passed
Add license for all SQL files........................................(no files to check)Skipped
Add license for all other files......................................(no files to check)Skipped
Add license for all rst files........................................(no files to check)Skipped
Add license for all JS/TS/TSX/CSS/PUML files.........................(no files to check)Skipped
Add license for all JINJA template files.............................(no files to check)Skipped
Add license for all shell files......................................(no files to check)Skipped
Add license for all Python files.........................................................Passed
Add license for all XML files........................................(no files to check)Skipped
Add license for all YAML files.......................................(no files to check)Skipped
Add license for all md files.........................................(no files to check)Skipped
Add license for all mermaid files....................................(no files to check)Skipped
Add TOC for md files.................................................(no files to check)Skipped
black....................................................................................Passed
blacken-docs.............................................................................Passed
Check for merge conflicts................................................................Passed
Debug Statements (Python)................................................................Passed
Check builtin type constructor use.......................................................Passed
Detect Private Key.......................................................................Passed
Fix End of Files.........................................................................Passed
Mixed line ending........................................................................Passed
Check that executables have shebangs.................................(no files to check)Skipped
Check Xml............................................................(no files to check)Skipped
Trim Trailing Whitespace.................................................................Passed
Fix python encoding pragma...............................................................Passed
Pretty format JSON...................................................(no files to check)Skipped
pyupgrade................................................................................Passed
rst ``code`` is two backticks........................................(no files to check)Skipped
use logger.warning(......................................................................Passed
Check YAML files with yamllint.......................................(no files to check)Skipped
Run isort to sort imports in Python files................................................Passed
Run pydocstyle...........................................................................Passed
Lint OpenAPI using spectral..........................................(no files to check)Skipped
Lint OpenAPI using openapi-spec-validator............................(no files to check)Skipped
Lint dockerfile......................................................(no files to check)Skipped
Check order of dependencies in setup.cfg and setup.py................(no files to check)Skipped
Checks setup extra packages..........................................(no files to check)Skipped
Update output of breeze command in BREEZE.rst........................(no files to check)Skipped
Update mounts in the local yml file..................................(no files to check)Skipped
Update setup.cfg file with all licenses..............................(no files to check)Skipped
Build cross-dependencies for providers packages......................(no files to check)Skipped
Update extras in documentation.......................................(no files to check)Skipped
Check for pydevd debug statements accidentally left......................................Passed
Don't use safe in templates..........................................(no files to check)Skipped
No providers imports in core example DAGs............................(no files to check)Skipped
No relative imports......................................................................Passed
Check for language that we do not accept as community....................................Passed
Check for inconsistent pylint disable/enable without space...............................Passed
Check BaseOperator[Link] core imports....................................................Passed
Check BaseOperator[Link] other imports...............................(no files to check)Skipped
Check provide_session and create_session imports.........................................Passed
Make sure LoggingMixin is not used alone.................................................Passed
Make sure days_ago is imported from airflow.utils.dates..................................Passed
'start_date' should not be defined in default_args in example_dags...(no files to check)Skipped
Check if integration list is aligned.................................(no files to check)Skipped
Check if licenses are OK for Apache..................................(no files to check)Skipped
Checks for consistency between config.yml and default_config.cfg.....(no files to check)Skipped
Verify that DB Migrations have been documented.......................(no files to check)Skipped
Checks for Boring Cyborg configuration consistency.......................................Passed
Sort INTHEWILD.md alphabetically.....................................(no files to check)Skipped
Sort alphabetically and uniquify spelling_wordlist.txt...............(no files to check)Skipped
Lint Helm Chart......................................................(no files to check)Skipped
Check Shell scripts syntax correctness...............................(no files to check)Skipped
Run BATS bash tests for changed Breeze bash files....................(no files to check)Skipped
Run BATS bash tests for changed bash files...........................(no files to check)Skipped
stylelint............................................................(no files to check)Skipped
Provider init file.......................................................................Passed
Validate providers.yaml files........................................(no files to check)Skipped
Generate mermaid images..............................................(no files to check)Skipped
Check if pre-commits are described...................................(no files to check)Skipped
Ensure hook ids are not overly long..................................(no files to check)Skipped
Checks providers available when declared by extras in setup.py.......(no files to check)Skipped
Run markdownlint.....................................................(no files to check)Skipped
Lint JSON Schema files with JSON Schema..............................(no files to check)Skipped
Lint NodePort Service with JSON Schema...............................(no files to check)Skipped
Lint Docker compose files with JSON Schema...........................(no files to check)Skipped
Lint chart/values.schema.json file with JSON Schema..................(no files to check)Skipped
Lint chart/values.yaml file with JSON Schema.........................(no files to check)Skipped
Lint airflow/config_templates/config.yml file with JSON Schema.......(no files to check)Skipped
Convert to f-strings with flynt..........................................................Passed
ESLint against airflow/ui............................................(no files to check)Skipped
ESLint against current UI js files...................................(no files to check)Skipped
Run in container bats tests..........................................(no files to check)Skipped
Check if image build is needed...........................................................Passed
Run mypy.................................................................................Passed
Run mypy for helm chart tests........................................(no files to check)Skipped
Run mypy for /docs/ folder...........................................(no files to check)Skipped
Run pylint for main code.................................................................Passed
Run pylint for tests.....................................................................Passed
Run pylint for helm chart tests......................................(no files to check)Skipped
Run flake8...............................................................................Passed
```

File: airflow/models/taskinstance.py
Comment (ecerulm) [reply airflow/models/taskinstance.py]: good catch, I'll update this when we are decided on `conn.value.conn_id` vs `conn.conn_id`

**Comment by ecerulm** [comment]
@potiuk  The CI job is failing due to pylint in a file I didn't touch: 

    tests/jobs/test_scheduler_job.py:3663:15: E1101: Instance of 'str' has no 'task_id' member (no-member)

I'm rebased to the latest `origin/branch`  commit 2833c2b2ffda682cc11893bc5efb1f53a254097e

**Comment by potiuk** [comment]
See #16682

File: airflow/models/taskinstance.py
Comment (ashb) [reply airflow/models/taskinstance.py]: So for variable is makes sense as there are two different ways of accessing a variable (as a plain string, or trying to json parse it) but I can't think of any case where we'd need this for connections, which always return "structured" data.

Have I missed something?

Comment (ecerulm) [reply airflow/models/taskinstance.py]: No , there is **no need**, it's just for having a **consistent api for both var and conns**. I mean not having to remember that when you access a variable you  need `var.value.xxx` but when you access a connection is just `conn.xxx`. But sure, I'll change it.

**Review by potiuk — APPROVED** [review]

**Comment by ecerulm** [comment]
I did change the connection accessor syntax from `conn.value.my_connection.host` to `conn.my_connection.host` as per request of @ashb.

Also, I guess I need to do a PR towards airflow-site too? https://github.com/apache/airflow/edit/main/docs/apache-airflow/macros-ref.rst , right?  Do I do that already now or we shall wait until this PR is merged.

**Comment by potiuk** [comment]
> Also, I guess I need to do a PR towards airflow-site too? https://github.com/apache/airflow/edit/main/docs/apache-airflow/macros-ref.rst , right? Do I do that already now or we shall wait until this PR is merged.

Good point. Not really PR to airflow-site - you need to update the https://github.com/apache/airflow/blob/main/docs/apache-airflow/macros-ref.rst  in the current PR. The docs at "airflow-site" are generated from there when we release.

**Comment by eladkal** [comment]
Question: how is it going to work for connection custom fields? for example: Asana has `workspace` in a separate connection field

**Comment by kaxil** [comment]
> workspace

That would be `{{ conn.a_connection.extra_dejson['extra__asana__workspace'] }}` but good point, we should add test for such case probably

**Review by kaxil — COMMENTED** [review]
We should add docs in https://airflow.apache.org/docs/apache-airflow/stable/concepts/connections.html (similar to https://airflow.apache.org/docs/apache-airflow/stable/concepts/variables.html) about getting connections in template.

And in https://airflow.apache.org/docs/apache-airflow/stable/macros-ref.html

The source code for those files are:

- https://github.com/apache/airflow/blob/main/docs/apache-airflow/macros-ref.rst
- https://github.com/apache/airflow/blob/main/docs/apache-airflow/concepts/connections.rst

**Comment by ecerulm** [comment]
Ok, I added testcase for `extra_dejson` and  update documentation in `macros-ref.rst` and `connnections.rst`

**Review by potiuk — APPROVED** [review]
Really nice now!

**Comment by potiuk** [comment]
Would you please rebase @ecerulm ? We have merged two stability improvements for CI (pylint + helm race condition).  It should help a a bit :)

File: tests/models/test_taskinstance.py
Comment (kurtqq) [inline tests/models/test_taskinstance.py:1571]: this test cover only connection defined in the database.
connection can also be in secret backend - will the macro work with it? shouldn't it be also covered in test?

Comment (potiuk) [reply tests/models/test_taskinstance.py:1571]: It's covered already by the unit tests of the methods used `get_connection_from_secrets`. There is no need to repeat those.

**Review by eladkal — APPROVED** [review]

**Review by kaxil — APPROVED** [review]

File: docs/apache-airflow/concepts/connections.rst
Comment (uranusjr) [inline docs/apache-airflow/concepts/connections.rst]: `connection_id`?

File: docs/apache-airflow/macros-ref.rst
Comment (uranusjr) [inline docs/apache-airflow/macros-ref.rst]: Is this a good example? If `my_conn_id` exists this is the same as `conn.my_conn_id`, and if it does not the `.host` part results in an error.

Comment (ecerulm) [reply docs/apache-airflow/macros-ref.rst]: This one just show that it's possible to fetch a conn_id via an string it's not about the defaults. I mean the `.get()` has two advantages one is that you can lookup a conn_id calculates at runtime like `conn.get('my_conn_'+index)`. The other advantage is that you can provide defaults. 

It the same examples for `var.value.get` if I recall properly I just copied them and and changed the wording.

Comment (uranusjr) [reply docs/apache-airflow/macros-ref.rst]: OK then

File: docs/apache-airflow/concepts/connections.rst
Comment (ecerulm) [reply docs/apache-airflow/concepts/connections.rst]: yes, fixed

**Comment by ashb** [comment]
> `{{ conn.a_connection.extra_dejson['extra__asana__workspace'] }}`

Because jinja is flexible this can also be done as

`{{ conn.a_connection.extra_dejson.extra__asana__workspace }}`

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]: Does a fallback make sense here? For variables is was easier as it is a "simple" object, but connection is much more complex, so I can't think _what_ we would put in the fallback value to make the template be useful?

(Not saying we should remove it, just that maybe the `'fallback'` isn't the best example here.

File: docs/apache-airflow/concepts/connections.rst
Comment (ashb) [inline docs/apache-airflow/concepts/connections.rst:30]:
```suggestion
    echo {{ conn.my_conn_id.host }}
```

File: airflow/models/taskinstance.py
Comment (ashb) [reply airflow/models/taskinstance.py]: Ah, you've got a useful example in the macros doc. Maybe just put a "see macros" here to avoid duplication.

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py:1664]: Is it worth doing something like:


```suggestion
                    if isinstance(default_conn, str):
                        return Connection(uri=default_conn)
                    else:
                        return default_conn
```

to support passing a URI as the fallback and have it parsed to a Connection object?

**Comment by ashb** [comment]
(None of my comments are blocking -- if you think they are useful go for it)

**Comment by ecerulm** [comment]
@potiuk the CI still fails for reason I believe unrelated to the content of this PR.  I can rebase again but judging from the last PR merged https://github.com/apache/airflow/pull/15397/checks  which had the same checks failing as this one, I don't think that will help.

**Comment by potiuk** [comment]
Yeah. It looks like a problem with the GitHub (public) runners this time. There seem to be no relation between Python/DB or any other consistent pattern with those failures. Maybe you could apply some of the comments from @ashb and when you do  - we merge it finally.

File: docs/apache-airflow/concepts/connections.rst
Comment (ecerulm) [reply docs/apache-airflow/concepts/connections.rst:30]: the document  for variables (variable.rst) uses the `echo {{ conn.<variable_name> }}` so I rather keep it this way to maintain consistency between those two.

File: airflow/models/taskinstance.py
Comment (ecerulm) [reply airflow/models/taskinstance.py:1664]: won't be that encouraging one of the security pitfalls ? Users then will write the URI in the dag source code and the URI most likely contains a secret, right?  Or did I misunderstood you

Comment (ashb) [reply airflow/models/taskinstance.py:1664]: Yeah, you're probably right.

**Review by ashb — APPROVED** [review]

**Review by uranusjr — APPROVED** [review]

**Comment by potiuk** [comment]
Something  fishy going on with helm installation on public runners randomly:

```
  Error: failed post-install: timed out waiting for the condition
```

**Comment by potiuk** [comment]
But unrelated. Merging.

**Pull request merged**
